src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -237,51 +256,68 @@ void limitStepOut (Quality const& ofrQ,
    If callback returns false, don't process any more offers.
    Return the unfunded and bad offers and the number of offers consumed.
 */
-template <class TAmtIn, class TAmtOut, class Callback>
-static
+template <class TIn, class TOut>
+template <class Callback>
 std::pair<boost::container::flat_set<uint256>, std::uint32_t>
-forEachOffer (
+BookStep<TIn, TOut>::forEachOffer (
     PaymentSandbox& sb,
     ApplyView& afView,
-    Book const& book,
-    AccountID const& src,
-    AccountID const& dst,
     bool prevStepRedeems,
-    bool ownerPaysTransferFee,
-    Callback& callback,
-    std::uint32_t limit,
-    beast::Journal j)
+    Callback& callback) const
 {
     // Charge the offer owner, not the sender
     // Charge a fee even if the owner is the same as the issuer
     // (the old code does not charge a fee)
     // Calculate amount that goes to the taker and the amount charged the offer owner
-    auto transferRate = [&](AccountID const& id)->std::uint32_t
+    auto transferRate = [this, &sb](AccountID const& id)->std::uint32_t
     {
-        if (isXRP (id) || id == dst)
+        if (isXRP (id) || id == this->strandDst_)
             return QUALITY_ONE;
         return rippleTransferRate (sb, id);
     };
 
     std::uint32_t const trIn =
-            prevStepRedeems ? transferRate (book.in.account) : QUALITY_ONE;
+        prevStepRedeems ? transferRate (book_.in.account) : QUALITY_ONE;
     // Always charge the transfer fee, even if the owner is the issuer
     std::uint32_t const trOut =
-        ownerPaysTransferFee ? transferRate (book.out.account) : QUALITY_ONE;
+        ownerPaysTransferFee_ ? transferRate (book_.out.account) : QUALITY_ONE;
+
+    typename FlowOfferStream<TIn, TOut>::StepCounter
+        counter (maxOffersToConsume_, j_);
 
-    typename FlowOfferStream<TAmtIn, TAmtOut>::StepCounter counter (limit, j);
-    FlowOfferStream<TAmtIn, TAmtOut> offers (
-        sb, afView, book, sb.parentCloseTime (), counter, j);
+    FlowOfferStream<TIn, TOut> offers (
+        sb, afView, book_, sb.parentCloseTime (), counter, j_);
 
     boost::optional<Quality> ofrQ;
     while (offers.step ())
     {
         auto& offer = offers.tip ();
+
+        // Note that offer.quality() returns a (non-optional) Quality.  So
+        // ofrQ is always safe to use below this point in the loop.
         if (!ofrQ)
             ofrQ = offer.quality ();
         else if (*ofrQ != offer.quality ())
             break;
 
+        // Special rule for offer crossing.  If:
+        //   a. We're offer crossing using default path (no autobridging), and
+        //   b. The offer's quality is at least as good as our quality, and
+        //   c. We're about to cross one of our own offers, then
+        //   c. Delete the old offer.
+        // See explanation in the testSelfCrossOffer1() unit test.
+        if (removeSelfCrossingOffers_ &&
+            ofrQ >= limitQuality_ &&
```

Comment:
I can explain these behaviors (not saying we should rely on them).

The `operator>=` is coming from `optional`.  At least in Boost 1.60 `optional` provides this implementation:
```
template<class T>
inline
bool operator >= ( optional<T> const& x, T const& y )
{ return !( x < y ) ; }
```
Since `Quality` _does_ define `operator<` it just works.  Not that we necessarily want to rely on this peculiarity of `optional`.

I'll revisit all of this when I convert `limitQuality_` to an `optional` named `limitSelfCrossingQuality_`, which is a great suggestion of yours.